### PR TITLE
Add flag management page

### PR DESCRIPTION
**Pull request by Vinnl**
This adds the ability for anyone with an `@mozilla.com` email address and for whom the `manage_flags` flag is enabled, to enable or disable a flag for everyone.

Submitting for consideration; not sure yet whether this is a good approach. In particular, this statement [on the Waffle site](https://waffle.readthedocs.io/en/stable/about/goals.html) is not too reassuring:

> Waffle is *not* designed to be secure, or be a replacement for permissions

Although the code also checks whether a user is authenticated and has an `@mozilla.com` address, I wonder if there's more we'd need to check for, e.g. if we could somehow attach [model permissions](https://www.django-rest-framework.org/api-guide/permissions/#djangomodelpermissions) on Waffle's flags model.

This also probably won't be the final approach, but could potentially satisfy our needs until the real solution is in place - perhaps even just enabling this when `RELAY_CHANNEL` is not `prod`.

### CODE REVIEW

File: frontend/src/pages/flags.page.tsx [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+              autoComplete="off"
+              required={true}
+              placeholder={`e.g. \`${flags[0]?.name}\``}
+              value={flagInput}
+              onChange={(e) => setFlagInput(e.target.value)}
+            />
+          </div>
+          <div className={styles.field}>
+            <label htmlFor="action">
+              Do you want to <output>enable</output> or <output>disable</output>{" "}
```
Comment (jwhitlock) [inline frontend/src/pages/flags.page.tsx:136]: I'd like a third option. Setting `everyone` to `true` or `false` turns it on or off globally, like a switch. `null` means there is no global option, and you can set or clear the flag at the user or group level or as a percent.

**Review by jwhitlock — COMMENTED** [review]
Thanks @Vinnl, it is good to see some progress on this feature!

I need to think a bit about the experience that we want, and the security model. This seems to work well for many of our workflows, where we turn features on and off in different environments. However, it seems as necessary to allow turning features on and off for a particular test user, some of which may have just been created, and leave the feature off for everyone else.

I added some suggested changes, and I'm thinking what we would need to get most of this merged safely and iterate.

File: api/views/__init__.py [hunk 2]
```diff
@@ -195,6 +196,46 @@ def runtime_data(request):
...
+        return flags
+
+    def perform_create(self, serializer):
+        if (
+            not flag_is_active(self.request, "manage_flags")
+            or not self.request.user.email.endswith("@mozilla.com")
+            or serializer.validated_data["name"] == "manage_flags"
+        ):
+            # TODO: Raise a proper 403 Forbidden error
+            raise RelayAPIException()
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply api/views/__init__.py]: Ah, thanks for the pointer, that is much cleaner indeed: eaa4dd4f

Would you also have a pointer on how to throw a 400 when someone tries to modify the `manage_flags` flag?

File: api/views/__init__.py [hunk 3]
```diff
@@ -195,6 +196,46 @@ def runtime_data(request):
...
+
+class FlagViewSet(SaveToRequestUser, viewsets.ModelViewSet):
+    serializer_class = FlagSerializer
+    permission_classes = [permissions.IsAuthenticated]
+    filterset_class = FlagFilter
+
+    def get_queryset(self):
+        if not flag_is_active(
+            self.request, "manage_flags"
+        ) or not self.request.user.email.endswith("@mozilla.com"):
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply api/views/__init__.py]: Ah sorry, the negative disjunction here is confusing, it should be clearer in eaa4dd4f. But not everyone at Mozilla would have the ability to modify flags; only those who have the `manage_flags` permission. So the idea would be to give that flag to a limited set of accounts, _and_ require that those accounts have `@mozilla.com` email addresses.

And since the API doesn't allow people to modify the `manage_flags` flag, only SRE would be able to give people that ability (so SREs would be "flag owners", and they can make people "flag managers"). There would be no "flag users" yet.

I'd be happy to have this managed by Django groups and permissions, which indeed sounds like the more appropriate tool for the job. And I think that might actually be what @say-yawn is looking at for the "real" solution, which would use the admin UI?

File: api/serializers/__init__.py [hunk 4]
```diff
@@ -151,6 +152,19 @@ class Meta:
         read_only_fields = ["email"]
 
 
+class FlagSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = get_waffle_flag_model()
+        fields = [
+            "id",
+            "name",
+            "everyone",
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply api/serializers/__init__.py:161]: I haven't made used of it in the UI, but that should at least be possible with e48c50be07814e5e1ef35cf8c1149fe9f77856cc.

File: frontend/src/pages/flags.page.tsx (hunk 1 above)
Comment (Vinnl) [reply frontend/src/pages/flags.page.tsx:154]: Hmm, so I started looking at this, but both the code as well as the UI got pretty complex pretty quickly, and then I realised that maybe this wouldn't be that useful anyway, since you can't currently set non-global flags.

What I did do, however, is to warn people when they try to modify a non-global flag, to avoid them discarding individual settings. If those have been set (i.e. by SRE), then from then on the flag can only be modified by SRE. bb542af1dea318d03fe98413f361fb9b969b88e9

**Comment by Vinnl** [comment]
Thanks @jwhitlock!

> However, it seems as necessary to allow turning features on and off for a particular test user, some of which may have just been created, and leave the feature off for everyone else.

Yes, this would absolutely be great. My intent with this PR is to offer a stopgap solution that would allow us (/me :smile:) to unblock QA on stage when SRE is not available, while awaiting the "real" solution that @say-yawn is working on - which _would_ allow granular modification of flags, but AFAIK requires alignment with other teams and is therefore not ready yet. When that becomes available, we can revert this PR again, as far as I'm concerned.

**Review by groovecoder — CHANGES_REQUESTED** [review]
change (blocking): we want to restrict the access to waffle flags to a smaller set of people than everyone with an @mozilla.com email address.

@say-yawn - what's the latest on putting the django admin UI behind LDAP SSO?

**Comment by Vinnl** [comment]
@groovecoder 

> change (blocking): we want to restrict the access to waffle flags to a smaller set of people than everyone with an @mozilla.com email address.

See https://github.com/mozilla/fx-private-relay/pull/2843#discussion_r1034554001 - it's not everyone with an `@mozilla.com` email address, it's everyone with an `@mozilla.com` email address _and_ with the `manage_flags` flag active for their user.

File: api/views/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwhitlock) [reply api/views/__init__.py]: Yes, add a `validate_name` to the serializer: https://www.django-rest-framework.org/api-guide/serializers/#field-level-validation

**Comment by jwhitlock** [comment]
> See https://github.com/mozilla/fx-private-relay/pull/2843#discussion_r1034554001 - it's not everyone with an @mozilla.com email address, it's everyone with an @mozilla.com email address and with the manage_flags flag active for their user.

Thanks for the clarification, I did not read the code closely enough. That makes me more comfortable about starting from here and building on it.

The restriction would be clearer with some unit tests, if that isn't too much of a stretch. If it is, we can test in round 2.

**Comment by Vinnl** [comment]
@jwhitlock I took a quick look this afternoon, but couldn't find what existing tests to model those after. If there are any and they're easy to find, could you point me to them? (I'll try to find them myself tomorrow, but I figured I'd ask in advane in case I don't manage.)

**Comment by jwhitlock** [comment]
> @jwhitlock I took a quick look this afternoon, but couldn't find what existing tests to model those after. If there are any and they're easy to find, could you point me to them? (I'll try to find them myself tomorrow, but I figured I'd ask in advane in case I don't manage.)

I couldn't find any either. I wrote some for PR #2805, but setup is a pain:

https://github.com/mozilla/fx-private-relay/blob/a1a7b80521dee521f8e8e9a915378e4e9a7cacba/api/tests/phones_views_tests.py

File: frontend/src/pages/flags.page.tsx (hunk 1 above)
Comment (jwhitlock) [reply frontend/src/pages/flags.page.tsx:154]: That makes sense to me. We can add more subtle option when we iterate.

File: api/views/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Vinnl) [reply api/views/__init__.py]: OK, now added proper validation with a 400 rejection in the HTTP method handlers, rather than in the serialiser: 3e1aad36

[the remark this answers is not in the document]
Comment (jwhitlock) [reply api/views/__init__.py]: Sorry I sent you down the wrong path 🤦 . We had a shared session to look at the code, here's what we discussed.

The `PATCH` method is by `id`, and doesn't include the `name` in the body, so it makes sense that `validate_name` isn't called. We could check if it is the `id` for the `manage_flags`, but it is probably better to wait until DRF loads the related model.

I think you're on the right path with `perform_create` and `perform_update`, when the model has been loaded. They are also not in the DRF documentation, so we're probably doing it wrong, and something could break on a DRF update. We can validate it still works with a test (next PR), and we can read the DRF code and docs to find the "right" place. For now, let's get it working and try it out.

**Comment by jwhitlock** [comment]
This worked for me:

* Remove `SaveToRequestUser` from `FlagViewSet`. That sets `user` to the `request.user`, and is not valid for the `Flag` model. This change allows creating a new flag from the web page.
* Remove `FlagViewSet.perform_create` and `FlagViewSet.perform_update`
* Add `FlagSerializer.validate()`:

```python
class FlagSerializer(serializers.ModelSerializer):
    class Meta:
        model = get_waffle_flag_model()
        fields = [
            "id",
            "name",
            "everyone",
            "note",
        ]
        read_only_fields = [
            "id",
        ]

    def validate(self, data):
        if (data.get("name", "").lower() == "manage_flags") or (
            hasattr(self, "instance")
            and getattr(self.instance, "name", "").lower() == "manage_flags"
        ):
            raise serializers.ValidationError(
                "Changing the `manage_flags` flag is not allowed."
            )
        return super().validate(data)
```

This looks for `name="manage_flags"` when creating a new flag, or an attempt to update the existing `manage_flags` instance.

File: api/views/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwhitlock) [reply api/views/__init__.py]: I'm wrong, they _are_ documented, I just couldn't find them with the doc search: https://www.django-rest-framework.org/api-guide/generic-views/

**Review by groovecoder — APPROVED** [review]
Approving so I'm not a blocking review anymore, but I'd like to make sure @jwhitlock is comfortable with the python code - I didn't check it in detail.

**Review by jwhitlock — APPROVED** [review]
> Approving so I'm not a blocking review anymore, but I'd like to make sure @jwhitlock is comfortable with the python code - I didn't check it in detail.

Yes, I'm still happy with the code. In the near future, I'd like to add some Python tests to duplicate my manual tests. I don't want to add too many more features, because we're thinking of a different solution in the future (see MPP-1922)
